File: ClassW/W1/1/dfa.py
class DFA:
    def __init__(self):
        self.alphabet = ['a', 'b']
        self.states = ['q0', 'q1', 'q2', 'q3', 'q4', 'q5', 'q6']
        self.transitions = {('q0', 'a'): 'q1', ('q0', 'b'): 'q2', ('q1', 'a'): 'q0', ('q1', 'b'): 'q5', ('q2', 'b'): 'q3', ('q3', 'b'): 'q4', ('q4', 'b'): 'q3', ('q5', 'b'): 'q6', ('q6', 'b'): 'q5'}
        self.start_state = 'q0'
        self.accepting_states = ['q0', 'q3', 'q5']

    def run(self, input_string):
        current_state = self.start_state
        for symbol in input_string:
            current_state = self.transitions.get((current_state, symbol))
            if current_state is None:
                return False
        return current_state in self.accepting_states

if __name__ == '__main__':
    dfa = DFA()
    print(dfa.run(input('Enter a string: ')))

File: ClassW/W1/1/dfa.cpp
#include <iostream>
#include <string>
using namespace std;

int main() {
    string input_string;
    cout << "Enter a string: ";
    cin >> input_string;

    int current_state = 0;
    for (char symbol : input_string) {
        switch (current_state) {
            case 0:
                if (symbol == 'a') current_state = 1;
                if (symbol == 'b') current_state = 2;
                break;
            case 1:
                if (symbol == 'a') current_state = 0;
                if (symbol == 'b') current_state = 5;
                break;
            case 2:
                if (symbol == 'b') current_state = 3;
                break;
            case 3:
                if (symbol == 'b') current_state = 4;
                break;
            case 4:
                if (symbol == 'b') current_state = 3;
                break;
            case 5:
                if (symbol == 'b') current_state = 6;
                break;
            case 6:
                if (symbol == 'b') current_state = 5;
                break;
        }
    }
    if (
        current_state == 0 ||
        current_state == 3 ||
        current_state == 5 
    ) {
        cout << "The string is accepted" << endl;
    } else {
        cout << "The string is not accepted" << endl;
    }
    return 0;
}